test/Dialect/FIRRTL/dedup.mlir

```diff
@@ -0,0 +1,411 @@
+firrtl.circuit "Empty" {
+  // CHECK: firrtl.module @Empty0
+  firrtl.module @Empty0(in %i0: !firrtl.uint<1>) { }
+  // CHECK-NOT: firrtl.module @Empty1
+  firrtl.module @Empty1(in %i1: !firrtl.uint<1>) { }
+  // CHECK-NOT: firrtl.module @Empty2
+  firrtl.module @Empty2(in %i2: !firrtl.uint<1>) { }
+  firrtl.module @Empty() {
+    // CHECK: %e0_i0 = firrtl.instance e0  @Empty0
+    // CHECK: %e1_i0 = firrtl.instance e1  @Empty0
+    // CHECK: %e2_i0 = firrtl.instance e2  @Empty0
+    %e0_i0 = firrtl.instance e0 @Empty0(in i0: !firrtl.uint<1>)
+    %e1_i1 = firrtl.instance e1 @Empty1(in i1: !firrtl.uint<1>)
+    %e2_i2 = firrtl.instance e2 @Empty2(in i2: !firrtl.uint<1>)
+  }
+}
+
+
+// CHECK-LABEL: firrtl.circuit "Simple"
+firrtl.circuit "Simple" {
+  // CHECK: firrtl.module @Simple0
+  firrtl.module @Simple0() {
+    %a = firrtl.wire: !firrtl.bundle<a: uint<1>>
+  }
+  // CHECK-NOT: firrtl.module @Simple1
+  firrtl.module @Simple1() {
+    %b = firrtl.wire: !firrtl.bundle<b: uint<1>>
+  }
+  firrtl.module @Simple() {
+    // CHECK: firrtl.instance simple0 @Simple0
+    // CHECK: firrtl.instance simple1 @Simple0
+    firrtl.instance simple0 @Simple0()
+    firrtl.instance simple1 @Simple1()
+  }
+}
+
+
+// CHECK-LABEL: firrtl.circuit "PrimOps"
+firrtl.circuit "PrimOps" {
+  // CHECK: firrtl.module @PrimOps0
+  firrtl.module @PrimOps0(in %a: !firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) {
+    %a_a = firrtl.subfield %a(0): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %a_b = firrtl.subfield %a(1): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %a_c = firrtl.subfield %a(2): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %0 = firrtl.xor %a_a, %a_b: (!firrtl.uint<2>, !firrtl.uint<2>) -> !firrtl.uint<2>
+    firrtl.connect %a_c, %a_b: !firrtl.uint<2>, !firrtl.uint<2>
+  }
+  // CHECK-NOT: firrtl.module @PrimOps1
+  firrtl.module @PrimOps1(in %b: !firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) {
+    %b_a = firrtl.subfield %b(0): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %b_b = firrtl.subfield %b(1): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %b_c = firrtl.subfield %b(2): (!firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>) -> !firrtl.uint<2>
+    %0 = firrtl.xor %b_a, %b_b: (!firrtl.uint<2>, !firrtl.uint<2>) -> !firrtl.uint<2>
+    firrtl.connect %b_c, %b_b: !firrtl.uint<2>, !firrtl.uint<2>
+  }
+  firrtl.module @PrimOps() {
+    // CHECK: firrtl.instance primops0 @PrimOps0
+    // CHECK: firrtl.instance primops1 @PrimOps0
+    %primops0 = firrtl.instance primops0 @PrimOps0(in a: !firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>)
+    %primops1 = firrtl.instance primops1 @PrimOps1(in b: !firrtl.bundle<a: uint<2>, b: uint<2>, c flip: uint<2>>)
+  }
+}
+
+// CHECK-LABEL: firrtl.circuit "Annotations"
+firrtl.circuit "Annotations" {
```

Comment:
This test hits on an angle that never occurred to me... When merging two modules that have multiple possible match points, it does bring up a question about if annotations should factor in to help disambiguate points which can match multiple ways. I.e., any of the wires could theoretically be deduplicated to any other wire and it is possible that a "lower energy" representation could be achieved by using common annotations as hints for deduplication. Don't do this as it complicates the problem substantially and I expect any benefit is balanced by the fact that this is an exceedingly rare occurrence.